### ISSUE TITLE
UseStatusCodePagesWithReExecute with UseStaticFiles results in wrong status code (200 OK)

### ISSUE THREAD

**Issue by Bouke** [dotnet/aspnetcore#40518]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

I have a few static files, one for each status code. I want to display those when an error code happens. I'm using a combination of `UseStatusCodePagesWithReExecute` and `UseStaticFiles`. However the `StaticFilesMiddleware` will replace the existing status code with 200, so now any error page is served with 200 OK instead of the actual status code.

```c#
app.UseStatusCodePagesWithReExecute("/Errors/{0}.html");
app.UseStaticFiles();
```

### Expected Behavior

The status code should reflect the actual error.

### Steps To Reproduce

```c#
var builder = WebApplication.CreateBuilder(args);
var app = builder.Build();
app.UseStatusCodePagesWithReExecute("/error.html");
app.UseStaticFiles();
app.MapGet("/", async context => { context.Response.StatusCode = 500; });
app.Run();
```

### .NET Version

6.0.200

**Comment by adityamandaleeka** [comment]
This is expected when you use `UseStatusCodePagesWithReExecute`. The status code you get back is from the re-executed request. You can capture the original status code yourself if you want when it's passed in to the path as '{0}' and set that as the status code on the response (something like this: https://stackoverflow.com/questions/46373668/when-using-usestatuscodepageswithreexecute-statuscode-is-not-sent-to-browser).

**Comment by Bouke** [comment]
@adityamandaleeka thank you for taking the time to respond. I understand that I can use a dynamically generated page to have the correct status code. My bug report is that I'm using `StaticFilesMiddleware`, which always returns 200. So does this mean that I cannot use `UseStatusCodePagesWithReExecute` with `UseStaticFiles` and have the correct response status code?

**Comment by adityamandaleeka** [comment]
You might be able to write middleware to do this. Unfortunately there's no built-in way to make UseStatusCodePagesWithReExecute preserve the original status code.

**Comment by Tratcher** [comment]
Reopening based on feedback from https://github.com/dotnet/AspNetCore.Docs/issues/26935.

We could improve this by saving the status code on IStatusCodeReExecuteFeature so that it is available in case the response generator overwrites it, like in StaticFile's OnPrepareResponse.
We could also avoid setting default response status codes (200) in static files and mvc.

**Comment by adityamandaleeka** [comment]
@amcasey PTAL

**Comment by amcasey** [comment]
@Tratcher 

> We could improve this by saving the status code on IStatusCodeReExecuteFeature so that it is available in case the response generator overwrites it, like in StaticFile's OnPrepareResponse. 

I managed to figure this out on my own, so nice to have confirmation. 😆 

> We could also avoid setting default response status codes (200) in static files and mvc.

I'm not sure what this means.  My guess was to not have the static files middleware clobber an existing status code, but the status code is non-nullable and defaults to 200.  Maybe you meant only change the status code if it's already 200?  What's the "and mvc" part?

**Comment by Tratcher** [comment]
> > We could also avoid setting default response status codes (200) in static files and mvc.
> 
> I'm not sure what this means. My guess was to not have the static files middleware clobber an existing status code, but the status code is non-nullable and defaults to 200. Maybe you meant only change the status code if it's already 200? What's the "and mvc" part?

The flow looks like this:
- The server creates the HttpContext with the default 200 status code.
- Some endpoint sets a status code (e.g. 401)
- UseStatusCodePagesWithReExecute reruns the pipeline
- StaticFiles doesn't look at the current status code, it always overwrites it.
https://github.com/dotnet/aspnetcore/blob/f58f9546be23e6480569b2143e7aec168f666b2a/src/Middleware/StaticFiles/src/StaticFileContext.cs#L243-L245

I'd meant something like this:
```C#
 public Task ApplyResponseHeadersAsync(int statusCode) 
 { 
     if (statusCode != 200)
     {
         _response.StatusCode = statusCode; 
```
But your suggestion might be better:
```C#
 public Task ApplyResponseHeadersAsync(int statusCode) 
 { 
     if (_response.StatusCode == 200)
     {
         _response.StatusCode = statusCode; 
```

**Comment by amcasey** [comment]
@Tratcher So you want to do that _in addition_ to exposing a feature property so that people can roll their own?  And what's the "and mvc" part?  Is there another place to apply a similar change?

**Comment by Tratcher** [comment]
Yes, we should do both. Make individual components behave better, but store the status code in the feature for when something doesn't. I'll have to find the MVC example, it's somewhere in ActionResults.

**Comment by Tratcher** [comment]
Here's one example, I'm sure there are others.
https://github.com/dotnet/aspnetcore/blob/f58f9546be23e6480569b2143e7aec168f666b2a/src/Mvc/Mvc.Core/src/ObjectResult.cs#L87
